kernel/src/engine/arrow_expression.rs

```diff
@@ -873,4 +900,206 @@ mod tests {
+            vec![Arc::new(StructArray::from(vec![
+                (
+                    Arc::new(Field::new("b", DataType::Int32, true)),
+                    create_array!(Int32, [1]) as ArrayRef,
+                ),
+                (
+                    Arc::new(Field::new("c", DataType::Int32, false)),
+                    create_array!(Int32, [2]) as ArrayRef,
+                ),
+            ]))],
+        )
+        .unwrap();
+        assert_create_one(values, schema, expected);
+    }
+
+    #[test]
+    fn test_create_one_nested_null() {
+        let values: &[Scalar] = &[Scalar::Null(DeltaDataTypes::INTEGER), 1.into()];
+        let schema = Arc::new(StructType::new([StructField::not_null(
+            "a",
+            DeltaDataTypes::struct_type([
+                StructField::nullable("b", DeltaDataTypes::INTEGER),
+                StructField::not_null("c", DeltaDataTypes::INTEGER),
+            ]),
+        )]));
+        let expected_schema = Arc::new(Schema::new(vec![Field::new(
+            "a",
+            DataType::Struct(
+                vec![
+                    Field::new("b", DataType::Int32, true),
+                    Field::new("c", DataType::Int32, false),
+                ]
+                .into(),
+            ),
+            false,
+        )]));
+        let expected = RecordBatch::try_new(
+            expected_schema,
+            vec![Arc::new(StructArray::from(vec![
+                (
+                    Arc::new(Field::new("b", DataType::Int32, true)),
+                    create_array!(Int32, [None]) as ArrayRef,
+                ),
+                (
+                    Arc::new(Field::new("c", DataType::Int32, false)),
+                    create_array!(Int32, [1]) as ArrayRef,
+                ),
+            ]))],
+        )
+        .unwrap();
+        assert_create_one(values, schema, expected);
+    }
+
+    #[test]
+    fn test_create_one_not_null_struct() {
+        let values: &[Scalar] = &[
+            Scalar::Null(DeltaDataTypes::INTEGER),
+            Scalar::Null(DeltaDataTypes::INTEGER),
+        ];
+        let schema = Arc::new(StructType::new([StructField::not_null(
+            "a",
+            DeltaDataTypes::struct_type([
+                StructField::not_null("b", DeltaDataTypes::INTEGER),
+                StructField::nullable("c", DeltaDataTypes::INTEGER),
+            ]),
+        )]));
+        let handler = ArrowExpressionHandler;
+        assert!(handler.create_one(schema, values).is_err());
+    }
+
+    #[test]
+    fn test_create_one_top_level_null() {
+        let values = &[Scalar::Null(DeltaDataTypes::INTEGER)];
+        let handler = ArrowExpressionHandler;
+
+        let schema = Arc::new(StructType::new([StructField::not_null(
+            "col_1",
+            DeltaDataTypes::INTEGER,
+        )]));
+        assert!(matches!(
+            handler.create_one(schema, values),
+            Err(Error::InvalidStructData(_))
+        ));
+    }
 }
```

Comment:
oof on the file size. We should probably split the tests out, and even split this all up. Maybe just make an issue for it (would be a good first issue)

Reply:
https://github.com/delta-io/delta-kernel-rs/issues/745

Reply:
done now and painful rebase haha